src/main/java/net/neoforged/neoforge/fluids/crafting/SingleFluidIngredient.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.fluids.crafting;
+
+import com.mojang.serialization.MapCodec;
+import java.util.stream.Stream;
+import net.minecraft.core.Holder;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.world.level.material.Fluid;
+import net.minecraft.world.level.material.Fluids;
+import net.neoforged.neoforge.common.NeoForgeMod;
+import net.neoforged.neoforge.fluids.FluidStack;
+import net.neoforged.neoforge.fluids.FluidType;
+
+/**
+ * Fluid ingredient that only matches the fluid of the given stack.
+ * <p>
+ * Unlike with ingredients, this is an explicit "type" of fluid ingredient,
+ * though it may still be written without a type field, see {@link FluidIngredient#MAP_CODEC}
+ */
+public class SingleFluidIngredient extends FluidIngredient {
+    public static final MapCodec<SingleFluidIngredient> CODEC = BuiltInRegistries.FLUID.holderByNameCodec()
```

Comment:
This should use `FluidStack.FLUID_NON_EMPTY_CODEC` (that can be made public since the corresponding `ItemStack` method is public) to disallow `Fluids.EMPTY`.